### PR TITLE
ICS 15: Cosmos Signed Messages

### PR DESCRIPTION
**Pull request by alexanderbez**
A (modified) port of the Cosmos Signed Message spec from the SDK. I initialized this spec as `draft` as it already had some form of approval in the SDK.

closes: #15

### CODE REVIEW

**Comment by cwgoes** [comment]
Do you mind using ICS 15? Joon & I assigned numbers to all of the issues - https://github.com/cosmos/ics/issues - so that PRs won't conflict.

**Comment by cwgoes** [comment]
Can we also note which functions we require from the signature algorithm (`sign` and `verify`), explain how they are used in the procedures, reference the JSON schema validation function (presumably there are libraries which do this)? Also ref https://github.com/cosmos/ics/issues/15#issuecomment-471541533 - ideally we should be able to encapsulate ASCII text in the JSON schema `@type` field.

**Comment by cwgoes** [comment]
For now, I think we can choose how to construct the "domain separator" based on accounts on Cosmos SDK-esque chains signing messages - so it should include account number / address, some kind of nonce, timestamp at minimum.

**Comment by cwgoes** [comment]
Please reformat re: https://github.com/cosmos/ics/issues/84 (in addition to prior comments)

**Comment by cwgoes** [comment]
Per https://github.com/cosmos/ics/issues/185#issuecomment-522240197, and the observation that this is the only non-IBC specification being considered for inclusion, I wonder if we should put this in the SDK repository (which I believe has other specifications for SDK modules) instead.

**Comment by cwgoes** [comment]
Per the above I'm going to close this. Feel free to reopen if you have objections.

I would be glad to review if this turns into an SDK-side PR.

**Comment by alexanderbez** [comment]
@cwgoes does this belong in ICS or the SDK? How does arbitrary signing of messages relate to IBC?

**Comment by cwgoes** [comment]
> @cwgoes does this belong in ICS or the SDK? How does arbitrary signing of messages relate to IBC?

It does not, and everything else in this repo is just for IBC, so I'm thinking this might be better off in the SDK.

File: spec/ics-15-cosmos-signed-messages/README.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "$id": "cosmos/signing/typeData/schema",
+  "title": "The Cosmos signed message typed data schema.",
+  "type": "object",
+  "properties": {
+    "@chain_id": {
+      "type": "string",
+      "description": "The corresponding Cosmos chain identifier.",
+      "minLength": 1
```
Comment (jleni) [inline spec/ics-15-cosmos-signed-messages/README.md]: Is it possible to specify `maxLength` for all (most?) fields?

**Comment by jleni** [comment]
- It would be good to restrict fields to ASCII or even 32..127 when possible to ensure that they can be represented in HW wallets.

- With respect to backwards compatibility, is there any specific requirement? It would be good to make that explicit.

- Canonical JSON: Make some restrictions explicit. Field order for instance (alphabetical or similar). Otherwise, identical content where field order changes results in different hashes / signatures. Similarly, spaces, separators, serialized objects (sets, etc.) that can result in multiple valid representations should be restricted to a unique representation.

- Field names: As a general recommendation, it would be good if field names, in particular in embedded objects:
1) could be user friendly so final users can understand them or...
2) there is some additional field with a short description that could be used for representational purposes

All in all, while this may require some changes in the HW wallet app but I don't see any significant impact that is not solvable. Actually, some improvements would result in a friendlier user interface.

**Pull request closed without merging**